File: App/Services/Notebook/NotebookConflict.swift
import Foundation

struct NotebookConflict: Decodable, Equatable {
    let conflict_id: UUID
    let base_version: Int64
    let server_version: Int64
    let incoming_title: String?
    let incoming_body: String?
    let server_title: String?
    let server_body: String?
}
struct NotebookConflictPage: Decodable {
    struct Detail: Decodable {
        let note_id: UUID; let title: String?; let body: String?; let version: Int64
        let tombstone: Bool; let updated_at: String; let conflicts: [NotebookConflict]
        var record: NotebookRecord { .init(note_id: note_id, title: title, body: body, version: version, tombstone: tombstone, updated_at: updated_at) }
    }
    let schema_version: Int; let note: Detail
    let has_more_conflicts: Bool; let next_conflict_after: UUID?
    func validate(noteID: UUID, after: UUID?) throws {
        try note.record.validate()
        guard schema_version == 1, note.note_id == noteID, note.conflicts.count <= 20,
              !note.tombstone || note.conflicts.isEmpty,
              has_more_conflicts ? note.conflicts.count == 20 && next_conflict_after == note.conflicts.last?.conflict_id : next_conflict_after == nil else { throw NotebookFailure.invalid }
        var previous = after?.uuidString ?? ""
        for conflict in note.conflicts {
            guard conflict.conflict_id.uuidString > previous, (0...9007199254740991).contains(conflict.base_version),
                  (1...note.version).contains(conflict.server_version),
                  [conflict.incoming_title, conflict.server_title].allSatisfy({ ($0?.unicodeScalars.count ?? 0) <= 200 }),
                  [conflict.incoming_body, conflict.server_body].allSatisfy({ ($0?.unicodeScalars.count ?? 0) <= 20000 }) else { throw NotebookFailure.invalid }
            previous = conflict.conflict_id.uuidString
        }
    }
}

File: android/core/data/src/main/kotlin/com/riskdetectedan/core/data/notebook/NotebookConflict.kt
package com.riskdetectedan.core.data.notebook

import kotlinx.serialization.Serializable
import java.util.UUID

@Serializable data class NotebookConflict(val conflict_id: String, val base_version: Long, val server_version: Long,
    val incoming_title: String?, val incoming_body: String?, val server_title: String?, val server_body: String?)
@Serializable data class NotebookConflictDetail(val note_id: String, val title: String?, val body: String?, val version: Long,
    val tombstone: Boolean, val updated_at: String, val conflicts: List<NotebookConflict>) {
    fun record() = NotebookRecord(note_id, title, body, version, tombstone, updated_at)
}
@Serializable data class NotebookConflictPage(val schema_version: Int, val note: NotebookConflictDetail,
    val has_more_conflicts: Boolean, val next_conflict_after: String?) {
    fun validate(noteID: String, after: String?) {
        note.record().validate()
        require(schema_version == 1 && note.note_id == noteID && note.conflicts.size <= 20 && (!note.tombstone || note.conflicts.isEmpty()))
        require(if (has_more_conflicts) note.conflicts.size == 20 && next_conflict_after == note.conflicts.last().conflict_id else next_conflict_after == null)
        var previous = after ?: ""
        for (conflict in note.conflicts) {
            require(UUID.fromString(conflict.conflict_id).toString() == conflict.conflict_id && conflict.conflict_id > previous)
            require(conflict.base_version in 0..9007199254740991L && conflict.server_version in 1..note.version)
            require(listOf(conflict.incoming_title, conflict.server_title).all { (it?.codePointCount(0, it.length) ?: 0) <= 200 })
            require(listOf(conflict.incoming_body, conflict.server_body).all { (it?.codePointCount(0, it.length) ?: 0) <= 20000 })
            previous = conflict.conflict_id
        }
    }
}
